From 0b581c144f5c56119bd30b832d4423fdc26435a8 Mon Sep 17 00:00:00 2001
From: Shammah Chancellor <shammah.chancellor@notvalid.com>
Date: Wed, 14 Oct 2026 16:42:07 -0700
Subject: [PATCH] tests: replace Bitcoin mining nonce fixtures with mined
 Dogecoin regtest blocks

Maintenance rationale (2026-10-14):
Execute the inherited mining validity scenarios with dynamically solved easy Scrypt headers and Dogecoin chain IDs instead of Bitcoin SHA256 nonce fixtures. The pinned Doged reference skipped these scenarios; this repair restores active coverage.

diff --git a/src/test/miner_tests.cpp b/src/test/miner_tests.cpp
index 40fcdf81914e283c12f6d2f9d8e1814446907380..749abe594dc22ff19c9b0368a1234a8a9d7b7347 100644
--- a/src/test/miner_tests.cpp
+++ b/src/test/miner_tests.cpp
@@ -17,6 +17,7 @@
 #include <consensus/validation.h>
 #include <node/blockfitter.h>
 #include <policy/policy.h>
+#include <pow/auxpow.h>
 #include <script/standard.h>
 #include <timedata.h>
 #include <txmempool.h>
@@ -58,7 +59,7 @@ struct MinerTestingSetup : public AvalancheTestingSetup {
     bool m_preconsensus{false};
 
     MinerTestingSetup(const std::vector<const char *> &extra_args = {})
-        : AvalancheTestingSetup(ChainType::MAIN, extra_args) {}
+        : AvalancheTestingSetup(ChainType::REGTEST, extra_args) {}
 
     void TestPackageSelection(const CChainParams &chainparams,
                               const CScript &scriptPubKey,
@@ -131,40 +132,6 @@ BlockAssembler MinerTestingSetup::AssemblerForTest(const CChainParams &params) {
         m_node.mempool.get(), assembler_options, m_node.avalanche.get()};
 }
 
-constexpr static struct {
-    uint8_t extranonce;
-    uint32_t nonce;
-} blockinfo[] = {
-    {4, 0xa4a3e223}, {2, 0x15c32f9e}, {1, 0x0375b547}, {1, 0x7004a8a5},
-    {2, 0xce440296}, {2, 0x52cfe198}, {1, 0x77a72cd0}, {2, 0xbb5d6f84},
-    {2, 0x83f30c2c}, {1, 0x48a73d5b}, {1, 0xef7dcd01}, {2, 0x6809c6c4},
-    {2, 0x0883ab3c}, {1, 0x087bbbe2}, {2, 0x2104a814}, {2, 0xdffb6daa},
-    {1, 0xee8a0a08}, {2, 0xba4237c1}, {1, 0xa70349dc}, {1, 0x344722bb},
-    {3, 0xd6294733}, {2, 0xec9f5c94}, {2, 0xca2fbc28}, {1, 0x6ba4f406},
-    {2, 0x015d4532}, {1, 0x6e119b7c}, {2, 0x43e8f314}, {2, 0x27962f38},
-    {2, 0xb571b51b}, {2, 0xb36bee23}, {2, 0xd17924a8}, {2, 0x6bc212d9},
-    {1, 0x630d4948}, {2, 0x9a4c4ebb}, {2, 0x554be537}, {1, 0xd63ddfc7},
-    {2, 0xa10acc11}, {1, 0x759a8363}, {2, 0xfb73090d}, {1, 0xe82c6a34},
-    {1, 0xe33e92d7}, {3, 0x658ef5cb}, {2, 0xba32ff22}, {5, 0x0227a10c},
-    {1, 0xa9a70155}, {5, 0xd096d809}, {1, 0x37176174}, {1, 0x830b8d0f},
-    {1, 0xc6e3910e}, {2, 0x823f3ca8}, {1, 0x99850849}, {1, 0x7521fb81},
-    {1, 0xaacaabab}, {1, 0xd645a2eb}, {5, 0x7aea1781}, {5, 0x9d6e4b78},
-    {1, 0x4ce90fd8}, {1, 0xabdc832d}, {6, 0x4a34f32a}, {2, 0xf2524c1c},
-    {2, 0x1bbeb08a}, {1, 0xad47f480}, {1, 0x9f026aeb}, {1, 0x15a95049},
-    {2, 0xd1cb95b2}, {2, 0xf84bbda5}, {1, 0x0fa62cd1}, {1, 0xe05f9169},
-    {1, 0x78d194a9}, {5, 0x3e38147b}, {5, 0x737ba0d4}, {1, 0x63378e10},
-    {1, 0x6d5f91cf}, {2, 0x88612eb8}, {2, 0xe9639484}, {1, 0xb7fabc9d},
-    {2, 0x19b01592}, {1, 0x5a90dd31}, {2, 0x5bd7e028}, {2, 0x94d00323},
-    {1, 0xa9b9c01a}, {1, 0x3a40de61}, {1, 0x56e7eec7}, {5, 0x859f7ef6},
-    {1, 0xfd8e5630}, {1, 0x2b0c9f7f}, {1, 0xba700e26}, {1, 0x7170a408},
-    {1, 0x70de86a8}, {1, 0x74d64cd5}, {1, 0x49e738a1}, {2, 0x6910b602},
-    {0, 0x643c565f}, {1, 0x54264b3f}, {2, 0x97ea6396}, {2, 0x55174459},
-    {2, 0x03e8779a}, {1, 0x98f34d8f}, {1, 0xc07b2b07}, {1, 0xdfe29668},
-    {1, 0x3141c7c1}, {1, 0xb3b595f4}, {1, 0x735abf08}, {5, 0x623bfbce},
-    {2, 0xd351e722}, {1, 0xf4ca48c9}, {1, 0x5b19c670}, {1, 0xa164bf0e},
-    {2, 0xbbbeb305}, {2, 0xfe1c810a},
-};
-
 static CBlockIndex CreateBlockIndex(int nHeight, CBlockIndex *active_chain_tip)
     EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
     CBlockIndex index;
@@ -215,7 +182,9 @@ void MinerTestingSetup::TestPackageSelection(
     // CTOR.
     // But using preconsensus, CTOR is always enforced.
     std::vector<TxId> txids = {mediumFeeTxId, parentTxId, highFeeTxId};
-    if (m_preconsensus) {
+    if (m_preconsensus || IsMagneticAnomalyEnabled(
+                                 chainparams.GetConsensus(),
+                                 m_node.chainman->ActiveTip())) {
         std::sort(txids.begin(), txids.end());
     }
     BOOST_CHECK(pblocktemplate->block.vtx[1]->GetId() == txids[0]);
@@ -761,7 +730,9 @@ void MinerTestingSetup::TestPrioritisedMining(
     std::vector<TxId> txids = {hashFreeParent, hashFreePrioritisedTx,
                                hashFreeChild, hashParentTx,
                                hashPrioritisedChild};
-    if (m_preconsensus) {
+    if (m_preconsensus || IsMagneticAnomalyEnabled(
+                                 chainparams.GetConsensus(),
+                                 m_node.chainman->ActiveTip())) {
         std::sort(txids.begin(), txids.end());
     }
     BOOST_CHECK(pblocktemplate->block.vtx[1]->GetId() == txids[0]);
@@ -781,8 +752,6 @@ void MinerTestingSetup::TestPrioritisedMining(
 
 // NOTE: These tests rely on CreateNewBlock doing its own self-validation!
 void MinerTestingSetup::CreateNewBlock_validity() {
-    // FIXME Update the below blocks to create a valid miner fund coinbase.
-    // This requires to update the blockinfo nonces.
     gArgs.ForceSetArg("-enableminerfund", "0");
     // Note that by default, these tests run with size accounting enabled.
     GlobalConfig config;
@@ -800,24 +769,18 @@ void MinerTestingSetup::CreateNewBlock_validity() {
 
     // We can't make transactions until we have inputs.
     // Therefore, load 110 blocks :)
-    static_assert(std::size(blockinfo) == 110,
-                  "Should have 110 blocks to import");
     int baseheight = 0;
     std::vector<CTransactionRef> txFirst;
-    for (const auto &bi : blockinfo) {
+    for (int height = 1; height <= 110; ++height) {
         // pointer for convenience
         CBlock *pblock = &pblocktemplate->block;
         {
             LOCK(cs_main);
-            pblock->nVersion = 1;
             pblock->nTime =
                 m_node.chainman->ActiveTip()->GetMedianTimePast() + 1;
             CMutableTransaction txCoinbase(*pblock->vtx[0]);
             txCoinbase.nVersion = 1;
-            txCoinbase.vin[0].scriptSig = CScript();
-            txCoinbase.vin[0].scriptSig.push_back(bi.extranonce);
-            txCoinbase.vin[0].scriptSig.push_back(
-                m_node.chainman->ActiveHeight());
+            txCoinbase.vin[0].scriptSig = CScript() << height << OP_0;
             txCoinbase.vout.resize(1);
             txCoinbase.vout[0].scriptPubKey = CScript();
             pblock->vtx[0] = MakeTransactionRef(std::move(txCoinbase));
@@ -828,12 +791,16 @@ void MinerTestingSetup::CreateNewBlock_validity() {
                 txFirst.push_back(pblock->vtx[0]);
             }
             pblock->hashMerkleRoot = BlockMerkleRoot(*pblock);
-            pblock->nNonce = bi.nonce;
+            pblock->nNonce = 0;
+            while (!CheckAuxProofOfWork(*pblock, chainparams.GetConsensus())) {
+                ++pblock->nNonce;
+                BOOST_REQUIRE(pblock->nNonce != 0);
+            }
         }
         std::shared_ptr<const CBlock> shared_pblock =
             std::make_shared<const CBlock>(*pblock);
-        BOOST_CHECK(Assert(m_node.chainman)
-                        ->ProcessNewBlock(shared_pblock, true, true, nullptr));
+        BOOST_REQUIRE(Assert(m_node.chainman)
+                          ->ProcessNewBlock(shared_pblock, true, true, nullptr));
         pblock->hashPrevBlock = pblock->GetHash();
     }
 
